pkg/queue/queue.go

```diff
@@ -135,80 +136,110 @@ func (q *RequestQueue) ReleaseRequests(items []Request) {
 func (q *RequestQueue) Dequeue(ctx context.Context, last QueueIndex, consumerID string) (Request, QueueIndex, error) {
+	dequeue, queueIndex, _, _, err := q.dequeue(ctx, last, anyQueueName, consumerID)
+	return dequeue, queueIndex, err
+}
+
+func (q *RequestQueue) dequeue(ctx context.Context, last QueueIndex, wantedQueueName string, consumerID string) (Request, QueueIndex, string, bool, error) {
 	q.mtx.Lock()
 	defer q.mtx.Unlock()
 
 	querierWait := false
 
 FindQueue:
 	// We need to wait if there are no tenants, or no pending requests for given querier.
-	for (q.queues.hasNoTenantQueues() || querierWait) && ctx.Err() == nil && !q.stopped {
+	// However, if `wantedQueueName` is not empty, the caller must not be blocked because it wants to read exactly from that queue, not others.
+	for (q.queues.hasNoTenantQueues() || querierWait) && ctx.Err() == nil && !q.stopped && wantedQueueName == anyQueueName {
 		querierWait = false
 		q.cond.Wait(ctx)
 	}
 
+	// If the current consumer wants to read from specific queue, but he does not have any queues available for him,
+	// return an error to notify that queue has been already removed.
+	if q.queues.hasNoTenantQueues() && wantedQueueName != anyQueueName {
+		return nil, last, wantedQueueName, false, ErrQueueWasRemoved
+	}
+
 	if q.stopped {
-		return nil, last, ErrStopped
+		return nil, last, wantedQueueName, false, ErrStopped
 	}
 
 	if err := ctx.Err(); err != nil {
-		return nil, last, err
+		return nil, last, wantedQueueName, false, err
 	}
 
 	for {
 		queue, tenant, idx := q.queues.getNextQueueForConsumer(last, consumerID)
 		last = idx
 		if queue == nil {
+			// it can be a case the consumer has other tenants queues available for him,
+			// it allows the consumer to pass the wait block,
+			// but the queue with index `last+1` has been already removed,
+			// for example if another consumer has read the last request from that queue,
+			// and as long as this consumer wants to read from specific tenant queue,
+			// it's necessary to return `ErrQueueWasRemoved` error.
+			if wantedQueueName != anyQueueName {
+				return nil, last, wantedQueueName, false, ErrQueueWasRemoved
+			}
+			// otherwise, if wantedQueueName is empty, then this consumer will go to the wait block again
+			// and as long as `last` index is updated, next time the consumer will request the queue
+			// with the new index that was returned from `getNextQueueForConsumer`.
 			break
 		}
 
+		if wantedQueueName != anyQueueName && wantedQueueName != queue.Name() {
+			// it means that the consumer received another tenants queue because it was already removed
+			// or another queue is already at this index
+			return nil, last, queue.Name(), false, ErrQueueWasRemoved
+		}
 		// Pick next request from the queue.
-		for {
-			request := queue.Dequeue()
-			if queue.Len() == 0 {
-				q.queues.deleteQueue(tenant)
-			}
+		request := queue.Dequeue()
+		tenantQueueEmpty := queue.Len() == 0
```

Comment:
nit: In Go, boolean variable names are mostly prefixed with `is/Is` or `has/Has` etc.

```suggestion
		isEmpty := queue.Len() == 0
```

Reply:
done, but I left it as `isTenantQueueEmpty` because I want to add additional meaning that not the whole queue is empty, but only tenant's queue